src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/segmentgroup/InvolvedParty.java

```diff
@@ -35,9 +35,9 @@
     public static final String INDICATOR = "S01";
 
     @Getter(lazy = true)
-    private final Optional<PerformingOrganisationNameAndAddress> organisationNameAndAddress =
-        extractOptionalSegment(PerformingOrganisationNameAndAddress.KEY_QUALIFIER)
-            .map(PerformingOrganisationNameAndAddress::fromString);
+    private final Optional<PerformerNameAndAddress> performerNameAndAddress =
```

Comment:
Just a question for yourself and @andrewwa-kainos.
I may have mentioned this before, but I believe the EDIFACT spec says that the `NAD` segment is required, so can we not safely assume that there will always be a performing organisation and a performer (same for a requesting organisation and requester), and use `extractSegment` here rather than `extractOptionalSegment`?

I feel we're over complicating this with the use of `Optional` everywhere

Reply:
That's also my understanding of the specification so I agree with you that we should use `extractSegment` to get the `NAD` segments as we should always have them for it to be a valid EDIFACT message.

Reply:
The way I read the document is that there must be exactly one `NAD` segment, which may be either `+PO` or `+SLA` (or `+MR`)
![image](https://user-images.githubusercontent.com/38751994/108981604-cbaff880-7684-11eb-921a-ed62357f9f72.png)
In this case, we're not guaranteed either of the specific classes - only the `NAD` key, for which we don't have an abstraction.

Reply:
If `NAD+PO` or `NAD+SLA` didn't exist, shouldn't we just set performer/requester to null (or something similar). Do we need all this `Optional`/`ifPresent` logic?

Reply:
I'd say that's for the consumer to decide. This class only knows that it might have one of these to return and it might not; which is what `Optional` is for. Returning `Optional` forces the consumer to consider the possibility it might be empty, instead of just assuming a value is given when it might actually be `null`.

Reply:
👍  That's fine, following the EDIFACT spec we can leave this for now